content/posts/notes-on-2pc.md

```diff
@@ -5,38 +5,73 @@ date: 2019-09-04
+### 2PC can be high available
+Any non-trivial protocol defines failure conditions that it's able to tolerate - 2PC is no exception. What is specific to 2PC is that some types of failures can make participants get "stuck". Whenever a participant votes "YES" it's unable to make any progress until hearing back from the coordinator. 
 
-### 2PC is the only commit protocol out-there
-No! E.g. FaunaDB.
+What are concrete reasons for getting stuck? First, the coordinator may fail. Secondly, the coordinator might be partitioned from the participant[^3]. The likelyhood of stucking is conditioned by coordinators availability and the probability of network partition. By making the failures less likely we can make 2PC more avaialble.
+
+{{< figure src="/posts/2pc-no-progress.jpg" title="Participant in the 'stuck' state">}}
+
+This touches on implementation and configuration argument already mentioned. For example MSDTC, the coordinator is a single process, however it can be deployed in the cluster environment which is a deployemnt decission. There also nothing in 2PC that prevents the coordinator to be implemened as a quorum of process[^4]. 
+
+Secondly, if all the parties (the coordinator and all participants) are running in the same local network, on a single cluster or inside a single VM, than what is the probability of network partitioning? As always, context is king.   
+
+### Commit latency is not the biggest problem
+Commiting in 2PC requires 2 round trips between coordinator and participants, and there are 4*n messages generated, where n is the number of participants. This is sometimes viewed as the root cause of many practical problems with the protocol. It defenatelly isn't ideal but it only surfaces other, bigger problem.
+
+The problem is potential contention at participant level caused by locking, especially in case of relational databases. Holding locks means that other transactions dealing with a given piece of state need to wait for the transaction to commit to make any progress.
+
+This behavior exists without 2PC but the protocol makes is pretty much always worst. In 2PC, the lock holding time is driven by the parital transation that takes the longes to commit, and the commit delay.
+
+### 2PC fits the cloud quite well
+We know that 2PC is used by the cloud vendors inside their services[^4] and can be used by the users when running at the level of IaaS[^5]. That said, none of the cloud vendors support MSDTC and/or XA at the level of native cloud services i.e. native service can't participante in 2PC. 
+
+Often, availability and performance are claimed to be the main reasons. Although significant, it can be argued that security is even more important. 2PC assumes high degree of trust between the participants and the coordinator. One could imagine an evil user tweaking the coordinator they own to exhausts participants resources by purposefully letting transactions hang in the `stuck state`. 
```

Comment:
```suggestion
Often times, availability and performance are claimed to be the reasons for that. Although these two are definitely the strongest points of 2PC, it can be argued that security (or its lack) is even more important. 2PC assumes high degree of trust between the participants and the coordinator. One could imagine an evil user that uses a specially crafted coordinator to exhausts participants resources by purposefully letting transactions hang in the `stuck state`. Similar attack could be performed by enlisting a malicious resource. 
```